### ISSUE TITLE
erroneous `note: erroneous constant used` emitted for non-compiling `format_args!` usage

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#110856]
I tried [this code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=cf1de58f32b3158d7ba97d7fd7363813):

```rust
format_args!("{:?}", { let [_a] = *(&[] as &[Vec<i32>]) else { return }; });
```

I expected to see this happen: an error about not being able to move out of a slice

Instead, this happened: the error and a bunch of `note: erroneous constant used`

```
error[E0508]: cannot move out of type `[Vec<i32>]`, a non-copy slice
 --> src/main.rs:2:35
  |
2 | format_args!("{:?}", { let [_a] = *(&[] as &[Vec<i32>]) else { return }; });
  |                             --    ^^^^^^^^^^^^^^^^^^^^^ cannot move out of here
  |                             |
  |                             data moved here
  |                             move occurs because `_a` has type `Vec<i32>`, which does not implement the `Copy` trait
  |
help: consider borrowing the pattern binding
  |
2 | format_args!("{:?}", { let [ref _a] = *(&[] as &[Vec<i32>]) else { return }; });
  |                             +++

note: erroneous constant used
 --> src/main.rs:2:14
  |
2 | format_args!("{:?}", { let [_a] = *(&[] as &[Vec<i32>]) else { return }; });
  |              ^^^^^^

note: erroneous constant used
 --> src/main.rs:2:37
  |
2 | format_args!("{:?}", { let [_a] = *(&[] as &[Vec<i32>]) else { return }; });
  |                                     ^^^

note: erroneous constant used
 --> src/main.rs:2:22
  |
2 | format_args!("{:?}", { let [_a] = *(&[] as &[Vec<i32>]) else { return }; });
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this note originates in the macro `format_args` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Meta

Fails on `1.71.0-nightly (2023-04-25 458d4dae845ec155b285)`, works fine on `1.70.0-beta.1 (2023-04-19 1b7dd2252b99671ce5d1)`

**Comment by Nilstrieb** [comment]
It also applies to way simpler code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=aeadbb4b94770923ec61da515e15e45c

**Comment by Nilstrieb** [comment]
searched nightlies: from nightly-2023-04-01 to nightly-2023-04-26
regressed nightly: nightly-2023-04-22
searched commit range: https://github.com/rust-lang/rust/compare/8bdcc62cb0362869f0e7b43a6ae4f96b953d3cbc...fec9adcdbc21469ef105162cc8cabf81c72d06be
regressed commit: https://github.com/rust-lang/rust/commit/fa4cc63a6bd6f32903269e52b8f59b284d0e76c4

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.6</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc 2023-04-01 --prompt --access github
```
</details>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc 2023-04-01 --prompt --access github
```
</details>

#110107 @cjgillot

Intestingly, the same day also contained two `format_args` PRs but I have bisected this twice (with manual promt) and const prop lint really does seem to be the cause:

commit[0] 2023-04-20: Auto merge of #110616 - m-ou-se:fmt-lang-items, r=jyn514
commit[1] 2023-04-20: Auto merge of #109999 - m-ou-se:flatten-format-args, r=oli-obk

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23110856.20erroneous.20.60note.3A.20erroneous.20constant.20used.60.20emitted.20f.E2.80.A6)).

@rustbot label -I-prioritize +P-medium
